banditpylib/learners/collaborative_learner/utils.py

```diff
@@ -0,0 +1,185 @@
+  def best_arm(self) -> int:
+    """Arm that the agent chose (could be None)"""
+
+  @abstractmethod
+  def broadcast(self) -> Dict[int, Tuple[float, int]]:
+    """Broadcasts information learnt in the current round
+
+    Returns: (a dict of)
+      * arm ids used in learning
+      * Tuple[corresponding average reward seen,
+        number of pulls used to deduce average]
+    """
+
+
+class CollaborativeBAIMaster(ABC):
+  r"""Master that handles arm assignment and elimination
+
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, name: Optional[str]):
+    self.__name = self._name() if name is None else name
+
+  @property
+  def name(self) -> str:
+    """Name of the master"""
+    return self.__name
+
+  @abstractmethod
+  def _name(self) -> str:
+    """
+    Returns:
+      default master name
+    """
+
+  @abstractmethod
+  def reset(self):
+    """Reset the master
+
+    .. warning::
+      This function should be called before the start of each game.
+    """
+
+  @abstractmethod
+  def initial_arm_assignment(self) -> Dict[int, List[int]]:
+    """The arm assignment for the first round
+
+    Returns:
+      dictionary of arm assignment per agent for all agents
+    """
+
+  @abstractmethod
+  def elimination(self, agent_ids: List[int],
+    messages: Dict[int, Tuple[float, int]]) -> Dict[int, List[int]]:
+    """Update the set of active arms based on some criteria
+    and return arm assignment
+
+    Args:
+      agent_ids: list of agents that will be assigned arms
+      messages: aggregation of messages broadcasted from agents
+
+    Returns:
+      dictionary of arm assignment per agent
+    """
+
+  @property
+  @abstractmethod
+  def active_arms(self):
+    """Arm indices that haven't been eliminated"""
+
+
+class CollaborativeBAILearner(Learner):
+  """Learner that puts the agents and master together
+
+  :param CollaborativeAgent agent: one instance of an agent
+  :param CollaboratveMaster master: instance of the master
+  :param int num_agents: total number of agents involved
+  :param Optional[str] name: alias name
+  """
+  def __init__(self, agent: CollaborativeBAIAgent,
+    master: CollaborativeBAIMaster, num_agents: int,
+    name: Optional[str] = None):
+    super().__init__(name)
+    self.__agents = []
+    for _ in range(num_agents):
+      self.__agents.append(dcopy(agent))
+    self.__master = master
+
+  def _name(self) -> str:
+    return 'collaborative_learner'
+
+  def reset(self):
+    for agent in self.__agents:
+      agent.reset()
+    self.__master.reset()
+
+  @property
+  def running_environment(self) -> Union[type, List[type]]:
+    return MultiArmedBandit
+
+  def get_agents(self) -> List[CollaborativeBAIAgent]:
+    """Involved agents"""
+    return self.__agents
+
+  def get_master(self) -> CollaborativeBAIMaster:
+    """Controlling Master"""
+    return self.__master
+
+  @property
+  def goal(self) -> Goal:
+    arm = Arm()
+    if len(self.__master.active_arms) == 1:
```

Comment:
Ask the learner for the best arm not the master.

Reply:
The algorithm says that the best arm is the only active arm or None

Reply:
Yes, I think it is talking about each agent. When the game stops, each agent is in state STOP. We should ask each agent her answer. If they are the same, the answer of the learner should be also the same. Otherwise, the learner has no answer.